File: year2024/day04/typescript/solution.ts
import { readFileSync } from 'fs';

const file: any = readFileSync('../' + (process.argv[2] === 'puzzle' ? 'puzzle' : 'example') + '.txt', 'utf-8');
const input: string[] = file.toString().trim().split('\n');
if(process.argv[2] === 'puzzle') {
  console.log(part_1());
  console.log(part_2());
}

export function part_1(): number {
  return input.reduce((count, row, r) => count + row.split('').reduce((innerCount, _, c) => innerCount + [[1, 1], [1, -1], [0, 1], [1, 0], [-1, -1], [-1, 1], [-1, 0], [0, -1]].reduce((acc, [dr, dc]) => acc + ([...Array(4)].every((_, i) => input[r + dr * i]?.[c + dc * i] === "XMAS"[i]) ? 1 : 0), 0), 0), 0);
}

export function part_2(): number {
  let [w, h, cw, res] = [input[0].length, input.length, word => word === 'SAM' || word === 'MAS', 0];
  for(let i = 1; i < w - 1; i++) for(let j = 1; j < h - 1; j++) {
    if(cw(`${input[i - 1][j - 1]}${input[i][j]}${input[i + 1][j + 1]}`) && cw(`${input[i + 1][j - 1]}${input[i][j]}${input[i - 1][j + 1]}`)) res++;
  }
  return res;
}

File: year2024/day04/python/solution.py
def main():
  print(part_1())
  print(part_2())


def part_1():
  n, m, res = len(input), len(input[0]), 0
  for i in range(n):
    for j in range(m):
      if i < n - 3 and input[i][j] + input[i + 1][j] + input[i + 2][j] + input[i + 3][j] in ("XMAS", "SAMX"):
        res += 1
      if j < n - 3 and input[i][j : j + 4] in ("XMAS", "SAMX"):
        res += 1
      if i < n - 3 and j < n - 3:
        if input[i][j] + input[i + 1][j + 1] + input[i + 2][j + 2] + input[i + 3][j + 3] in ("XMAS", "SAMX"):
          res += 1
        if input[i + 3][j] + input[i + 2][j + 1] + input[i + 1][j + 2] + input[i][j + 3] in ("XMAS", "SAMX"):
          res += 1
  return res


def part_2():
  return sum(input[i][j] + input[i + 1][j + 1] + input[i + 2][j + 2] in ("MAS", "SAM") and input[i + 2][j] + input[i + 1][j + 1] + input[i][j + 2] in ("MAS", "SAM") for i in range(len(input) - 2) for j in range(len(input[0]) - 2))


def setup(path):
  global input

  with open(path, 'r') as file:
    input = [row.strip() for row in file]


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()